### PR TITLE
restore: support table partition

### PR DESCRIPTION
**Pull request by 5kbpers**
Support table partition:
1. add rewrite rule for table partition
2. fix table partition checksum
3. refactor logic about key rewrite

### CODE REVIEW

File: tests/br_table_partition/run.sh [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    fi
+    echo "database $DB${i} [original] row count: ${row_count_ori[i]}, [after br] row count: ${row_count_new[i]}"
+done
+
+if $fail; then
+    echo "TEST: [$TEST_NAME] failed!"
+    exit 1
+else
+    echo "TEST: [$TEST_NAME] successed!"
+fi
```
Comment (overvenus) [inline tests/br_table_partition/run.sh]: Should `DROP DATABASE` before finishing the test.

File: pkg/restore/util.go [hunk 2]
```diff
@@ -237,94 +268,42 @@ func ValidateFileRewriteRule(file *backup.File, rewriteRules *restore_util.Rewri
...
-) *import_sstpb.RewriteRule {
-	for _, rule := range rewriteRules.Data {
-		// regions may have the new prefix
-		if bytes.HasPrefix(region.GetStartKey(), rule.GetNewKeyPrefix()) {
-			return rule
-		}
+// Rewrites a raw key and returns a encoded key
+func rewriteRawKey(key []byte, rewriteRules *restore_util.RewriteRules) ([]byte, *import_sstpb.RewriteRule) {
+	if len(key) > 0 {
+		rule := matchOldPrefix(key, rewriteRules)
```
Comment (3pointer) [inline pkg/restore/util.go]: should check nil here?

File: pkg/restore/util.go [hunk 3]
```diff
@@ -71,32 +77,48 @@ func GetRewriteRules(
...
-
-	dataRules := make([]*import_sstpb.RewriteRule, 0, len(oldTable.Indices)+1)
-	dataRules = append(dataRules, &import_sstpb.RewriteRule{
-		OldKeyPrefix: append(tablecodec.EncodeTablePrefix(oldTable.ID), recordPrefixSep...),
-		NewKeyPrefix: append(tablecodec.EncodeTablePrefix(newTable.ID), recordPrefixSep...),
-		NewTimestamp: newTimeStamp,
-	})
-
+	tableIDs := make(map[int64]int64)
+	tableIDs[oldTable.ID] = newTable.ID
```
Comment (3pointer) [inline pkg/restore/util.go]: if we got `oldTable.Partition != nil` then tableID is useless
should it be?
```
if oldTable.Partition != nil {
	for _, srcPart := range oldTable.Partition.Definitions {
		for _, destPart := range newTable.Partition.Definitions {
			if srcPart.Name == destPart.Name {
				tableIDs[srcPart.ID] = destPart.ID
			}
		}
	}
} else {
	tableIDs[oldTable.ID] = newTable.ID
}
```

Comment (5kbpers) [reply pkg/restore/util.go]: We don't have partitioned index so tableID is necessary.
BTW I think extra rewrite rules do not have a bad effect.

File: pkg/restore/client.go [hunk 4]
```diff
@@ -200,11 +198,6 @@ func (rc *Client) CreateTables(
 		Data:  make([]*import_sstpb.RewriteRule, 0),
 	}
 	newTables := make([]*model.TableInfo, 0, len(tables))
-	// Sort the tables by id for ensuring the new tables has same id ordering as the old tables.
-	// We require this constrain since newTableID of tableID+1 must be not bigger
-	// than newTableID of tableID.
-	sort.Sort(utils.Tables(tables))
```
Comment (overvenus) [inline pkg/restore/client.go]: Can we remove `utils.Table`?

File: pkg/restore/client.go [hunk 5]
```diff
@@ -215,21 +208,10 @@ func (rc *Client) CreateTables(
 			return nil, nil, err
 		}
 		rules := GetRewriteRules(newTableInfo, table.Schema, newTS)
-		tableIDMap[table.Schema.ID] = newTableInfo.ID
 		rewriteRules.Table = append(rewriteRules.Table, rules.Table...)
 		rewriteRules.Data = append(rewriteRules.Data, rules.Data...)
 		newTables = append(newTables, newTableInfo)
 	}
-	// If tableID + 1 has already exist, then we don't need to add a new rewrite rule for it.
-	for oldID, newID := range tableIDMap {
```
Comment (overvenus) [inline pkg/restore/client.go]: We should update validate.go since we change the rewrite process.

File: pkg/restore/util.go (hunk 3 above)
Comment (3pointer) [reply pkg/restore/util.go]: yes, extra rewrite rules seems no bad effect

**Pull request merged**